nimsuggest/nimsuggest.nim

```diff
@@ -7,6 +7,13 @@
 #    distribution, for details about the copyright.
 #
 
+import compiler/renderer
+import strformat
+import tables
+import std/sha1
+import segfaults
```

Comment:
The idea here is to catch segfaults as part of incremental recompilation and then do a clean compilation. In some of the future commits I will change to code to send the notification to the client about that failure to make sure that these crashes are properly fixed. Note that in any other case the uncaught exception will trigger that same behavior - the application will exit.

Reply:
To elaborate a bit more - I am trying to make nimsuggest more crash resilient to eliminate the need for having an external process wrapper(i. e. `nimlangserver`).

Reply:
I am also planning to replace `quit` calls in the compiler with throw exception (in case `nimsuggest`) to make crashes less likely. Please share your thoughts.

Reply:
The compiler keeps a tremendous amount of state around. This is pretty bad already and not easily fixed. If you continue after a bug the state becomes "corrupt", not necessarily in the sense that it causes a crash. A crash can easily be detected by nimlangserver or the like and the service can be restarted. If you instead make crashes disappear and at the same time nimsuggest produces wrong output, missing output etc then it becomes less trustworthy. It's not a good idea, Erlang's "let it crash" philosophy is the right idea and it relies on process isolation.

The process isolation is crucial for reliability. What we can do is to make nimsuggest a supervisor process like nimlangserver so that editor support does not have to reimplement the supervisor logic. Nimsuggest would then call something like a nimsuggest_worker process.

Reply:
> A crash can easily be detected by nimlangserver or the like and the service can be restarted. 

I had that same conversation with @zah - the problem here is that there is no easy way to detect where the crash comes from - if the crash comes from something else and nimlangserver restarts nimsuggest then the result might be a loop. 

> The process isolation is crucial for reliability. What we can do is to make nimsuggest a supervisor process like nimlangserver so that editor support does not have to reimplement the supervisor logic. Nimsuggest would then call something like a nimsuggest_worker process.

Isn't throwing away ModuleGraph equivalent to starting a clean process? My understanding is that there is no state outside of ModuleGraph. Or you have other concerns? FWIW I have tried to emulate creating "fresh" module graph in   `initModuleGraphFields` which will be called after recompile crashes. If we remove `segfaults` then we should remove the catch block here:  https://github.com/yyoncho/Nim/blob/nimsuggest-version-3/nimsuggest/nimsuggest.nim#L707 because it no longer makes sense if the strategy is `let it crash`. The java language server has a similar strategy in a slightly different context - it performs incremental builds as you type but in addition it has a command `Clean build` which is for the cases when the server context breaks up.

If you don't buy these arguments what do you think about having this behavior behind a build flag?

PS: I see this code evolving and this "patch" to me something that will go away once we address more and more recompilaton-related bugs. I have identified a few but I wanted to come up with something as usable as possible and then focus on fixing them.

Reply:
>  throw exception 

generally, in a compiler api built for tooling, exceptions are not the way to go in many / most cases - instead you want to collect errors along the way without aborting the process, meaning that you pass in some collector - if we're removing `quit` and the like, this is likely a better way to go

Reply:
@arnetheduck afaics that is what is happening in practice, it is just that when the collected error kind is in `fatalMsgs` it quits (https://github.com/yyoncho/Nim/blob/nimsuggest-version-3/compiler/lineinfos.nim#L214 ). So it is either throw an exception or let it continue. I guess for nimsuggest both should be good enough as long as it does not crash the process.